### PR TITLE
Request to merge methods-branch with develop branch

### PR DESCRIPTION
**Pull request by defoebrand**
This branch included creating the following methods in such a way as that they behave identically to their counterparts without the 'my_' prefix:

    my_each
    my_each_with_index
    my_select
    my_all?
    my_any?
    my_none?
    my_count
    my_map
    my_inject
As well as a 'multiply_els' method as a proof of 'my_inject' functionality.

A test file (test.rb) has been created with example tests taken from Ruby Docs. All tests are run once using default methods and again using the 'my_' methods. All results are identical.

### CODE REVIEW

**Review by trekab — CHANGES_REQUESTED** [review]
## Changes Required! :recycle: 
Hi team! Nice job so far. There are some additions/changes you need to make to your methods to have them behave as expected. Check out the points below:

- [x] `my_each` and `my_each_with_index` should return an Enumerator if no block is given.
- [x] For `my_each` and `my_each_with_index` when a block is given when self is a range, they should return the range itself after calling the given block once for each element in self.
- [x] `my_each` when a block is given when self is a hash behaves like `each` calls the given block once for each element in self.
- [x] `my_each_with_index` when a block is given when self is a hash behaves like `each_with_index` calls the given block once for each element in self
- [x] `my_select` should return an enumerator if no block is given.
- [x] `my_all?` should return false if the block returns false or nil::range
- [x] `my_all?` when no block or argument is given returns true when none of the collection members are false or nil.
- [x] `my_count` should return the number of items in enum through enumeration::range.
- [x] `my_inject` should raise a "LocalJumpError" when no block or argument is given.
- [x] `my_inject` should not mutate the original array
- [x] `my_inject` when a block is given without an initial value, it should combine all elements of the enum by applying a binary operation, specified by the block.
- [x] `my_inject` when a symbol is specified without an initial value, it should combine each element of the collection by applying the symbol as a named method:range.

__PS:__ Please try to check out the above behaviour in the [documentation](https://ruby-doc.org/core-2.6.5/Enumerable.html) and compare keenly to the current behaviour of your methods.

__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink: 
__Treasure Kabareebe__
[GitHub](https://github.com/trekab) | [LinkedIn](https://www.linkedin.com/in/treasure-kabareebe/) | [Twitter](https://twitter.com/TKabareebe)

**Comment by defoebrand** [comment]
All issues have been tested and adjusted to my understanding.

**Comment by lekegitrepo** [comment]
#### Hi @defoebrand and @Godswilly :clap: :clap:
#### Good job so far :+1: Bases on the previous review the following methods :point_down:  did not meet the requirement specified in by the previous reviewer.

- [ ] **_`my_each` and my_each_with_index should return an Enumerator if no block is given._**
- [ ] **_`my_select` should return an enumerator if no block is given._**
- [ ] **_`my_inject` when a block is given without an initial value, it should combine all elements of the enum by applying a binary operation, specified by the block, for example: `(5..10).my_inject { |sum, n| sum + n }` or `[2, 3, 5, 6, 1, 7, 5, 3, 9].my_inject { |sum, n| sum + n }`_**

**_fix these and submit them for review._**

#### _Reviewed by Temitayo Adeleke._

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
#### Status: Request changes

**Comment by defoebrand** [comment]
Both my_each enumerables, and my_select, return enumerators if no block is given
my_inject has been adjusted per reviewer suggestions
all methods have been updated with more descriptive names

**Review by mikenath223 — CHANGES_REQUESTED** [review]
### Hi Team. 👋

Really nice work so far. 👍

Let's fix this:

 - [ ] `my_each` and `my_each_with_index`  when a range is passed is expected to return a range after calling the block passed on each element in the range. However, yours returns an array instead of a range.

- [ ]  `my_each` and `my_each_with_index` when a hash is passed should call the block given on each element in the hash and return the result similar to what an `each method` would give.

- [ ]  `my_select` when a range is given and a block is passed is expected to return an array containing the elements for which the block passed is true but yours returns a hash instead. It should return an array.

- [ ] `my_all` throws an error when a range is passed in a case like this `range.my_all?(&false_block)`. The method should be built to receive a range.

Kindly fix these issues and request another review👍
Feel free to contact me on slack(michgolden) if you need further clarification.

HAPPY CODING👨‍💻

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+def t_f_test(arg, &block)
+  var = block.call(arg)
+  return true unless var.nil? || var == false
+end
+
+module Enumerable
+  def my_each(&block)
+    engine_select_block_check(&block)
+    self
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :014 > array.each.is_a?(Enumerator)
 => true 
2.6.5 :015 > array.my_each.is_a?(Enumerator)
 => false 
2.6.5 :016 > array
 => [3, 1, 8, 0, 0, 8, 4, 1, 0, 3, 2, 0, 2, 6, 8, 0, 4, 0, 8, 6, 8, 2, 3, 7, 1, 5, 3, 6, 5, 2, 7, 6, 0, 4, 5, 0, 6, 3, 8, 7, 5, 3, 4, 0, 8, 2, 3, 5, 3, 1, 0, 3, 7, 3, 2, 1, 2, 3, 7, 1, 6, 5, 1, 7, 0, 5, 6, 4, 2, 5, 0, 3, 8, 6, 3, 2, 3, 3, 2, 6, 6, 6, 0, 5, 3, 3, 2, 0, 4, 7, 0, 4, 0, 3, 8, 4, 3, 8, 3, 6] 
2.6.5 :017 >
```

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+  def my_each_with_index(_arg = nil, &block)
+    my_each(&block)
+    self
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :020 > array
 => [3, 1, 8, 0, 0, 8, 4, 1, 0, 3, 2, 0, 2, 6, 8, 0, 4, 0, 8, 6, 8, 2, 3, 7, 1, 5, 3, 6, 5, 2, 7, 6, 0, 4, 5, 0, 6, 3, 8, 7, 5, 3, 4, 0, 8, 2, 3, 5, 3, 1, 0, 3, 7, 3, 2, 1, 2, 3, 7, 1, 6, 5, 1, 7, 0, 5, 6, 4, 2, 5, 0, 3, 8, 6, 3, 2, 3, 3, 2, 6, 6, 6, 0, 5, 3, 3, 2, 0, 4, 7, 0, 4, 0, 3, 8, 4, 3, 8, 3, 6] 
2.6.5 :021 > array.each_with_index.is_a?(Enumerator)
 => true 
2.6.5 :022 > array.my_each_with_index.is_a?(Enumerator)
 => false 
2.6.5 :023 > 
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Thanks for applying the changes from the last reviews.

A bit more effort to move forward.
Please, read the comments left below and then apply the suggested changes.
Then submit a new code review request.

Thanks

Mael FOSSO

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  return true unless var.nil? || var == false
+end
+
+module Enumerable
+  def my_each(&block)
+    return to_enum unless block_given?
+
+    engine_select_block_check(&block)
+    self
+  end
```
Comment (christianotieno) [inline enumerables.rb:76]: # my_each
1. [ ] When a block is given when self is a hash behaves like #each calls the given block once for each element in self.
### Test 👉 `expect(my_each_output).to eq(each_output)`
|Expected|Got|
|:---:|:---:|
|` "[:a, 1][:b, 2][:c, 3][:d, 4][:e, 5]"`|`"abcde"`|

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+  def my_each_with_index(_arg = nil, &block)
+    return to_enum unless block_given?
+
+    my_each(&block)
+    self
+  end
```
Comment (christianotieno) [inline enumerables.rb]: # my_each_with_index 
1. [ ] When a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self
### Test 👉 ` expect(my_each_with_index_output).to eq(each_with_index_output)`
|Expected|Got|
|:---:|:---:|
|`"Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"`|`"Num: a, idx: 1\nNum: b, idx: 2\nNum: c, idx: 3\nNum: d, idx: 4\nNum: e, idx: 5\n"`|

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    array_query = []
+    if self.class == Hash
+      0.upto(length - 1) do |ind|
+        hash_query[keys[ind]] = values[ind] unless block.call(keys[ind], values[ind]) == false
+      end
+    else
+      engine_select_block_check { |val| array_query << val if t_f_test(val, &block) }
+    end
+    array_query.empty? ? hash_query : array_query
+  end
```
Comment (christianotieno) [inline enumerables.rb]: # my_select 
1. [ ] Returns an array containing all elements of enum for which the given block returns a true value::range
### Test 👉 `expect(range.my_select(&block)).to eq(range.select(&block))`
|Expected|Got|
|:---:|:---:|
|` []`|`{}`|

**Review by christianotieno — CHANGES_REQUESTED** [review]
# **CHANGES REQUIRED** :warning:
Hey @defoebrand & @lekegitrepo :wave:
-
You have done a splendid job with the changes! :tada: Keep it up.
We, however, do have some few more changes to carry out before we proceed (see review comment).
##
## NOTE: These are the variables I used in the tests below: 👇 
```
ARRAY_SIZE = 100
LOWEST_VALUE = 0
HIGHEST_VALUE = 9

describe 'enumerables' do
  let(:array) { Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } }
  let(:block) { proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } }
  let(:words) { %w[dog door rod blade] }
  let(:range) { Range.new(5, 50) }
  let(:hash) { { a: 1, b: 2, c: 3, d: 4, e: 5 } }
  let(:numbers) { [1, 2i, 3.14] }
  let!(:array_clone) { array.clone }
```
## Tests Done: 79
## Failures: 3
- If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email. :eyes:
##
- When you have made the specified changes, please submit another code review. Thank you!
Happy coding :smile:
**[Christian Otieno](https://github.com/ChristianOtieno)**
[Twitter](https://twitter.com/iamchrisotieno) [linkedin](https://www.linkedin.com/in/christianotieno/) [GitLab](https://gitlab.com/christianotieno)
---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

**Review by rOluochKe — APPROVED** [review]
Hello Team
### Well done, you got it right this time :clap: The project has been approved. Congratulations :boom: :100:
### Merge and submit the project. Good luck with the next one.
- Happy coding!

**Pull request merged**